src/py/mat3ra/made/tools/build/defect/builders.py

```diff
@@ -371,3 +378,143 @@ def _generate(self, configuration: _ConfigurationType) -> List[_GeneratedItemTyp
             thickness=configuration.thickness,
             use_cartesian_coordinates=configuration.use_cartesian_coordinates,
         )
+
+
+class TerraceSlabDefectBuilder(SlabDefectBuilder):
+    _ConfigurationType: type(TerraceSlabDefectConfiguration) = TerraceSlabDefectConfiguration  # type: ignore
+    _GeneratedItemType: Material = Material
+
+    def _calculate_cut_direction_vector(self, material, cut_direction):
+        np_cut_direction = np.array(cut_direction)
+        direction_vector = np.dot(np.array(material.basis.cell.vectors_as_nested_array), np_cut_direction)
+        normalized_direction_vector = direction_vector / np.linalg.norm(direction_vector)
+        return normalized_direction_vector
+
+    def _calcualte_height(self, material, new_material):
+        original_max_z = get_atomic_coordinates_extremum(material, use_cartesian_coordinates=True)
+        added_layers_max_z = get_atomic_coordinates_extremum(new_material, use_cartesian_coordinates=True)
+        height_cart = added_layers_max_z - original_max_z
+        return height_cart
+
+    def _calculate_rotation_parameters(self, original_material, new_material, normalized_direction_vector):
+        """Calculate the necessary rotation angle and axis."""
+        height_cart = self._calcualte_height(original_material, new_material)
+        cut_direction_xy_proj_cart = np.linalg.norm(
+            np.dot(np.array(new_material.basis.cell.vectors_as_nested_array), normalized_direction_vector)
+        )
+        # Slope of the terrace along the cut direction
+        hypotenuse = np.linalg.norm([height_cart, cut_direction_xy_proj_cart])
+        angle = np.arctan(height_cart / cut_direction_xy_proj_cart) * 180 / np.pi
+        normalized_rotation_axis = np.cross(normalized_direction_vector, [0, 0, 1]).tolist()
+        delta_length = hypotenuse - cut_direction_xy_proj_cart
+        return angle, normalized_rotation_axis, delta_length
+
+    def _increase_lattice_size(self, material, length_increase, direction_of_increase):
+        """Adjust the lattice vectors to accommodate the new structure dimensions."""
```

Comment:
Not sure what this does - needs a better explanation and an example in docstring maybe